File: src/1813.c
#include <math.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

bool areSentencesSimilar(char* sentence1, char* sentence2) {
    bool retVal = false;

#define STRTOK_PATTERN (" ")
    char* p = NULL;

    char* words1[strlen(sentence1)];
    int words1Size = 0;
    p = strtok(sentence1, STRTOK_PATTERN);
    while (p != NULL) {
        words1[words1Size++] = p;
        p = strtok(NULL, STRTOK_PATTERN);
    }

    char* words2[strlen(sentence2)];
    int words2Size = 0;
    p = strtok(sentence2, STRTOK_PATTERN);
    while (p != NULL) {
        words2[words2Size++] = p;
        p = strtok(NULL, STRTOK_PATTERN);
    }

    int head = 0;
    while ((head < words1Size) && (head < words2Size)) {
        if (strcmp(words1[head], words2[head]) != 0) {
            break;
        }
        ++head;
    }

    int tail = 0;
    while ((tail < words1Size - head) && (tail < words2Size - head)) {
        if (strcmp(words1[words1Size - 1 - tail], words2[words2Size - 1 - tail]) != 0) {
            break;
        }
        ++tail;
    }

    if ((head + tail) == fmin(words1Size, words2Size)) {
        retVal = true;
    }

    return retVal;
}

int main(int argc, char** argv) {
#define MAX_SIZE (100)
    struct testCaseType {
        char sentence1[MAX_SIZE];
        char sentence2[MAX_SIZE];
    } testCase[] = {{"My name is Haley", "My Haley"},
                    {"of", "A lot of words"},
                    {"Eating right now", "Eating"},
                    {"Luky", "Lucccky"}};
    int numberOfTestCase = sizeof(testCase) / sizeof(testCase[0]);
    /* Example
     *  Input: sentence1 = "My name is Haley", sentence2 = "My Haley"
     *  Output: true
     *
     *  Input: sentence1 = "of", sentence2 = "A lot of words"
     *  Output: false
     *
     *  Input: sentence1 = "Eating right now", sentence2 = "Eating"
     *  Output: true
     *
     *  Input: sentence1 = "Luky", sentence2 = "Lucccky"
     *  Output: false
     */

    bool answer = false;
    int i;
    for (i = 0; i < numberOfTestCase; ++i) {
        printf("Input: sentence1 = \"%s\",  sentence2 = \"%s\"\n", testCase[i].sentence1, testCase[i].sentence2);

        answer = areSentencesSimilar(testCase[i].sentence1, testCase[i].sentence2);
        printf("Output: %s\n", (answer == true ? "true" : "false"));

        printf("\n");
    }

    return EXIT_SUCCESS;
}

File: src/1813.py
import logging
import sys
from bisect import bisect_left, bisect_right
from collections import Counter, defaultdict, deque
from functools import cache, lru_cache
from heapq import heapify, heappop, heappush
from itertools import accumulate
from typing import List, Optional


def logging_setting():
    LOG_LEVEL = logging.INFO    # DEBUG, INFO, WARNING, ERROR, CRITICAL
    if __debug__:
        LOG_LEVEL = logging.DEBUG

    logging.basicConfig(
        stream=sys.stderr,
        level=LOG_LEVEL,
        format="%(levelname)-6s %(asctime)s %(module)s:%(funcName)s:%(lineno)-4s %(message)s",
        datefmt='%Y/%m/%d %H:%M:%S')


class Solution:
    def areSentencesSimilar(self, sentence1: str, sentence2: str) -> bool:
        retVal = False

        words1 = list(sentence1.split())
        words2 = list(sentence2.split())

        while words1 and words2 and (words1[0] == words2[0]):
            words1.pop(0)
            words2.pop(0)

        while words1 and words2 and (words1[-1] == words2[-1]):
            words1.pop(-1)
            words2.pop(-1)

        if (not words1) or (not words2):
            retVal = True

        return retVal


if __name__ == "__main__":
    logging_setting()

    try:
        logging.info("sys.version: %s", sys.version)
        print()

        pSolution = Solution()
        for sentence1, sentence2 in zip(["My name is Haley", "of", "Eating right now", "Luky"],
                                        ["My Haley", "A lot of words", "Eating", "Lucccky"]):
            # /* Example
            #  *  Input: sentence1 = "My name is Haley", sentence2 = "My Haley"
            #  *  Output: true
            #  *
            #  *  Input: sentence1 = "of", sentence2 = "A lot of words"
            #  *  Output: false
            #  *
            #  *  Input: sentence1 = "Eating right now", sentence2 = "Eating"
            #  *  Output: true
            #  *
            #  *  Input: sentence1 = "Luky", sentence2 = "Lucccky"
            #  *  Output: false
            #  */
            logging.info("Input: sentence1 = \"%s\", sentence2 = \"%s\"", sentence1, sentence2)

            retVal = pSolution.areSentencesSimilar(sentence1, sentence2)
            logging.info("Output: %s", retVal)

            print()
    except KeyboardInterrupt as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass
    except Exception as exception:
        logging.error("%s: %s", exception.__class__.__name__,
                      exception, exc_info=True)
        pass

    sys.exit(0)
